src/socket/socket_router.rs

```diff
@@ -655,19 +657,64 @@ impl SocketRouter {
     }
 
     async fn on_achievement_complete(&mut self, data: PlayerAchievementData) -> Result<(), SocketError> {
-        // confirm the achievement exists
-        let achievement_exists = Database::find_by_id(&self.server.api_state.database.achievements, data.achievement_id.as_str()).await.is_some();
-        if !achievement_exists {
+        // Confirm the achievement exists and retrieve it
+        let achievement_id = data.achievement_id.as_str();
+        let database = &self.server.api_state.database;
+        let collection = &database.achievements;
+
+        if let Some(mut achievement) = Database::find_by_id(collection, achievement_id).await {
+            // Check if first_completion is None
+            println!("\n\nID: {}\n\n", achievement.first_completion);
+            if Uuid::parse_str(&*achievement.first_completion).unwrap().is_nil() {
+                // Set first_completion to the current PlayerAchievementData
+                achievement.first_completion = data.player.id;
+
+                // Manually handle the update operation and log any errors
```

Comment:
Talked about this earlier but there is an existing function to save the updated achievement record to the database without having to do the manual serialization & update logic here